bpo-30362 : Add list options to launcher.

GadgetSteve opened the pull request:
These changes add to the windows launcher the options to list the available pythons, (-l or --list), or list with the paths, (-L or --long-list), and then terminate.

In the process the output of the help messages has been moved to a distinct function, (show_help_text).

Note that informational messages are sent to stderr and only the possible -x.y-bb flags or the flags and paths are output to stdout, this is intended to allow pipe of the flags for onward processing or scripting.

GadgetSteve committed:
bpo-30362 Add list options to launcher.

GadgetSteve committed:
bpo-30362 Add list options to help message.

GadgetSteve committed:
To avoid possible later conflict with python replaced flags with --launcher-list and --launcher-list-paths

vsajip commented:
I agree this is a useful feature, but I'm not sure I like the patch stylistically. Haven't yet had time to look at it in more detail.

GadgetSteve committed:
bpo-30362 Changed flag to -0 as suggested on review.

eryksun commented on PC/launcher.c:
```diff
@@ -1371,6 +1371,79 @@ get_version_info(wchar_t * version_text, size_t size)
+    else
+    {
+        for (i = 0; i < num_installed_pythons; i++, ip++) {
+            fwprintf(stdout, fmt, ip->version, ip->bits, ip->executable);
```
Does it really need to print the executable path? It's trivial to get that information manually.

eryksun commented on PC/launcher.c:
```diff
@@ -1371,6 +1371,79 @@ get_version_info(wchar_t * version_text, size_t size)
+    if (ip == NULL)
+        fwprintf(stderr, L"\n\nCan't find a Default Python.\n\n");
+    else
+        fwprintf(stderr, \
```
Instead of printing the default version to stderr, it could indicate the default version in the list with an asterisk. For example, `*3.6-64`.

GadgetSteve committed:
bpo-30362: Modified to default to not path for -0, -0p to dispaly path and append * to default

GadgetSteve committed:
bpo-30362: Modified to display list on required version not found.

GadgetSteve committed:
bpo-30362 add --list and --list-paths added back in following review by paul.moore

zooba commented on PC/launcher.c:
```diff
@@ -1516,12 +1598,25 @@ process(int argc, wchar_t ** argv)
+        if (argc == 2) {
+            slen = wcslen(L"-0");
+            if(!wcsncmp(p, L"-0", slen)) /* Starts with -0 */
+                show_python_list(argv); /* Check for -0 FIRST */
```
Can we return the exit code from `show_python_list` rather than using `exit()` in it? It's *probably* safe enough (because I'm sure enough people have done it in the past that the MS CRT protects itself against people doing it), but better to return from the entry point than to force an exit.

GadgetSteve replied:
Restructured to allow exit via return from main rather than direct exit call.

GadgetSteve committed:
bpo-30362 Cleaner handing of -0 & -0p by not calling exit directly per review by @zooba

zooba commented on PC/launcher.c:
```diff
@@ -1431,26 +1431,26 @@ show_python_list(wchar_t ** argv)
     }
 
-    ip = locate_python(L"", FALSE);
+    /*ip = locate_python(L"", FALSE);*/
```
Remove this if it's not needed now.

zooba commented:
Please also see https://devguide.python.org/committing/#what-s-new-and-news-entries and add a NEWS and what's new item for this.

GadgetSteve committed:
bpo-30362: Tidy up and add news & what's new

Removed commented out line of code in PC/launcher.c.
Added the results of using blurb to add details of bpo-30362 & bpo-30291.
Updated Doc/whatsnew/3.7.rst to add a Windows only section covering both tickets.

GadgetSteve committed:
bpo-30362 Resolve conflict in Doc/whatsnew/3.7.rst

GadgetSteve commented:
I am unable to resolve conflict in Doc/whatsnew/3.7.rst

GadgetSteve committed:
Merge branch 'master' into fix-issue-30362

GadgetSteve committed:
bpo-30362:Address Whitespace Issue in Doc\whatsnew\3.7.rst

GadgetSteve commented:
Conflict resolved.

zooba committed:
Shorten NEWS message for bpo-30362

zooba committed:
Shorten NEWS item for bpo-30291

zooba commented:
Thanks. I shortened the two NEWS items so they are in line with the rest of our items. Once the checks are complete then I'll merge. Thanks for this!
